src/NServiceBus.Core/Recoverability/DefaultRecoverabilityPolicy.cs

```diff
@@ -18,9 +18,13 @@ public static class DefaultRecoverabilityPolicy
         /// <returns>The recoverability action.</returns>
         public static RecoverabilityAction Invoke(RecoverabilityConfig config, ErrorContext errorContext)
         {
-            if (errorContext.Exception is MessageDeserializationException)
+            // ReSharper disable once LoopCanBeConvertedToQuery
+            foreach (var unrecoverableExceptionType in config.Failed.UnrecoverableExceptionTypes)
```

Comment:
My original comment wasn't really about perf concerns. It was just the enumeration of the HashSet vs calling Contains that led me to raise the question about if considering inheritance here made sense.

Given how exceptions do tend to be inherited, I can see that it is beneficial here. If I added `ArgumentException` to the collection, it seems reasonable that `ArgumentNullException` would also be included.